File: tiles.jl
tic()

using PyPlot

numColours = 4
numRows = 4
numColumns = 2

function getRowCol(index)
    return (div(index-1,numColumns)+1, (index-1)%numColumns+1)
end

function calcDist(x, y)
    x1, x2 = getRowCol(x)
    y1, y2 = getRowCol(y)
    return sqrt((y1-x1)^2 + (y2-x2)^2)
end

function calcCost(group)
    return minimum([calcDist(z[1], z[2]) for z in filter(a->a[1]<a[2], [(x,y) for x in group, y in group])])
end

n = iceil(numRows*numColumns/numColours)
cellVals = repmat(1:numColours, n, 1)
cellVals = cellVals[1:numRows*numColumns]

allGroupings = Set()
for ordering in permutations(cellVals)
    curGroupings = Array{Int,1}[]
    for i in 1:numColours
        push!(curGroupings, filter(j->ordering[j]==i, 1:length(ordering)))
    end
    push!(allGroupings, curGroupings)
end

allGroupings = collect(allGroupings)

allCosts = [minimum([calcCost(colourGroup) for colourGroup in groupings]) for groupings in allGroupings]

maxGroups = allGroupings[indmax(allCosts)]

tiles = zeros(Int, numRows, numColumns)

curVal = 0
for colour in maxGroups
    for point in colour
        x,y = getRowCol(point)
        tiles[x,y] = curVal
    end
    curVal += 1
end

imshow(tiles, interpolation="nearest")
axis("off")

toc()

File: tiles.py
import numpy as np
import matplotlib.pyplot as plt
import math
import itertools
import time

t = time.time()


numColours = 4
numRows = 3
numColumns = 3


def getRowCol(index):
    return (index/numColumns, index%numColumns)

def calcDist(x, y):
    x1, x2 = getRowCol(x)
    y1, y2 = getRowCol(y)
    return math.sqrt((y1-x1)**2 + (y2-x2)**2)

def calcCost(group):
    return min([calcDist(x, y) for x in group for y in group if y>x])


n = int(math.ceil(float(numRows*numColumns)/numColours))
cellVals = n*range(numColours)
cellVals = cellVals[:numRows*numColumns]

allGroupings = set()
for ordering in itertools.permutations(cellVals):
    curGroupings = []
    for i in range(numColours):
        curGroupings.append(frozenset([j for j in range(len(ordering)) if ordering[j]==i]))
    allGroupings.add(frozenset(curGroupings))

allGroupings = list(allGroupings)
allCosts = [min([calcCost(colourGroup) for colourGroup in groupings]) for groupings in allGroupings]

maxGroups = list([allGroupings[i] for i in range(len(allGroupings)) if allCosts[i]==max(allCosts)][0])

tiles = np.zeros((numRows, numColumns))

curVal = 0
for colour in maxGroups:
    for point in colour: 
        tiles[getRowCol(point)] = curVal
    curVal += 1

plt.imshow(tiles, interpolation='nearest')
plt.axis('off')
plt.show()

elapsed = time.time() - t
print elapsed
